### PR TITLE
Loki tests

### PR DESCRIPTION
**Pull request by BeaudanBrown**
Set up the framework for loki tests, added some tests for libloki-protocol and proof-of-work

### CODE REVIEW

File: libloki/proof-of-work.js [hunk 1]
```diff
@@ -2,9 +2,17 @@ const hash = require('js-sha512');
...
+  incrementNonce,
+  bufferToBase64,
+  bigIntToUint8Array,
+  greaterThan,
+};
+
 const NONCE_LEN = 8;
 // Modify this value for difficulty scaling
-let NONCE_TRIALS = 1000;
+let NONCE_TRIALS = 10;
```
Comment (Mikunj) [inline libloki/proof-of-work.js]: intended change?

Comment (sachaaaaa) [inline libloki/proof-of-work.js]: I would prefer having both production and development constants defined up here:
```js
let NONCE_TRIALS = 1000;
let DEV_NONCE_TRIALS = 10;
```
and then below
```js
let nonce_trial = development ? DEV_NONCE_TRIAL : NONCE_TRIAL;
```

How does that sound?

File: libloki/test/libloki-protocol_test.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      1
+    );
+    fallbackCipher = new libloki.FallBackSessionCipher(address);
+  });
+
+  it('should encrypt fallback cipher messages as friend requests', async () => {
+    const buffer = new ArrayBuffer(10);
+    const { type } = await fallbackCipher.encrypt(buffer);
+    assert.strictEqual(type, textsecure.protobuf.Envelope.Type.FRIEND_REQUEST);
+  });
```
Comment (sachaaaaa) [inline libloki/test/libloki-protocol_test.js]: Would be good to actually test if the encryption worked by decrypting the message and checking if it matches the initial buffer.

File: libloki/test/libloki-protocol_test.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  beforeEach(async () => {
+    clearDatabase();
+    testKey = {
+      pubKey: libsignal.crypto.getRandomBytes(33),
+      privKey: libsignal.crypto.getRandomBytes(32),
+    };
+    await store.storeSignedPreKey(1, testKey);
+  });
+
+  it('should should generate a new prekey bundle for a new contact', async () => {
```
Comment (sachaaaaa) [inline libloki/test/libloki-protocol_test.js]:
```suggestion
  it('should generate a new prekey bundle for a new contact', async () => {
```

File: libloki/test/libloki-protocol_test.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    assert.isDefined(newBundle.signedKeyId);
+    assert.isDefined(newBundle.preKey);
+    assert.isDefined(newBundle.signedKey);
+    assert.isDefined(newBundle.signature);
+    assert.strictEqual(testKeyArray.byteLength, newBundle.signedKey.byteLength);
+    for (let i = 0 ; i !== testKeyArray.byteLength; i += 1)
+      assert.strictEqual(testKeyArray[i], newBundle.signedKey[i]);
+  });
+
+  it('should should return the same prekey bundle after creating a contact', async () => {
```
Comment (sachaaaaa) [inline libloki/test/libloki-protocol_test.js]:
```suggestion
  it('should return the same prekey bundle after creating a contact', async () => {
```

File: libloki/test/libloki-protocol_test.js (hunk 2 above)
Comment (BeaudanBrown) [reply libloki/test/libloki-protocol_test.js]: Whoops lol thought I did that

File: libloki/proof-of-work.js (hunk 1 above)
Comment (Mikunj) [reply libloki/proof-of-work.js]: should be `const` btw :p

Comment (BeaudanBrown) [reply libloki/proof-of-work.js]: Hm can you think of how to do that with testing each function individually? Cause that's how it was done before but that check happened what you send an ipc message
I couldn't think of a clean way to do it so just set the default to dev level but I agree it feels gross

**Pull request merged**